functionary/builder/utils.py

```diff
@@ -105,6 +255,109 @@ def initiate_build(
     return build
 
 
+def create_package_from_definition(
+    package_definition: dict, environment: "Environment"
+) -> Package:
+    """Create a package from a package definition"""
+    # TODO: Manually parsing for now, but this should be codified somewhere, with
+    #       the parsing informed by package_definition_version.
+    package_obj = Package(
+        environment=environment,
+        name=package_definition.get("name"),
+    )
+
+    package_obj.display_name = package_definition.get("display_name")
+    package_obj.summary = package_definition.get("summary")
+    package_obj.description = package_definition.get("description")
+    package_obj.language = package_definition.get("language")
+
+    package_obj.save()
+
+    return package_obj
+
+
+@app.task
+def build_package(build_id: "UUID") -> None:
+    """Retrieve the resources for Build and use them to build and push the package
+    docker image. Also creates BuildLog with build/push information.
+
+    Args:
+        build_id: ID of the build being executed
+
+    Returns:
+        None
+    """
+    build = Build.objects.select_related("environment").get(id=build_id)
+    build_resource = BuildResource.objects.get(build=build)
+
+    logger.info(f"Starting build {build.id}")
+
+    try:
+        build, log = _build_package(build, build_resource)
+        build.status = Build.COMPLETE
+        _create_build_log(build, log)
+        _update_build_status(build)
+        logger.info(f"Build {build_id} COMPLETE")
+    except BuilderException as err:
+        logger.error(f"Encountered build error: {err}")
+        build.status = Build.ERROR
+        _create_build_log(build, err)
+        _update_build_status(build)
+    except DockerSocketConnectionError as err:
+        logger.fatal(f"{err}")
+        raise err
+    except Exception as err:
+        logger.error(f"Build {build.id} encountered unexpected error: {err}")
+        build.status = Build.ERROR
+        _create_build_log(build, err)
+        _update_build_status(build)
+
+
+def _build_package(build: Build, build_resource: BuildResource) -> tuple[Build, str]:
+    """Generate and build necessary resources to complete image build
+
+    Args:
+        build: The build object containing metadata about the build
+
+    Returns:
+        build: The build object with updated attribute values
+        build_log: The log string of the entire build process
+
+    Raises:
+        DockerSockerConnectionError: Raised when connection to the docker socket
+        could not be established.
+    """
+    docker_client = _get_docker_client()
+
+    build.status = Build.IN_PROGRESS
+    _update_build_status(build)
+
+    # Set variables necessary for build process
+    build_log = ""
+    workdir = _generate_path_for_build(build)
+    package: Package = build.package
```

Comment:
Since I am guaranteeing that the package exists before the `build_image` method runs, I include the `Package` annotation

Reply:
Would it not be better to update the model to be accurate?  It sounds like with your changes the expectation is that `build.package` should never be `None`.  If you update the field on the model to reflect that, then you don't need the annotation here because the type of `build.package` is actually guaranteed at that point.